enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,249 @@
+  A new custom deploy method `install_assisted` will be added to start the
+  assisted agent at the right moment.
+- The assisted service will stop explicitly disabling *inspection* on
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+N/A (this change itself does not add new user stories)
+
+### API Extensions
+
+All required API extensions are already present.
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
+`InfraEnv` resource. In both cases it will generate an Ignition configuration
+to start and configure the Ironic agent. In the ZTP case, the corresponding
+`InfraEnv` will be updated with the generated Ignition configuration (merging
+it with any user-provided configuration).
+
+The same process will be applied to `initrd` images for (i)PXE booting. The
+Ignition configuration will be embedded into the resulting `initrd` by the
+assisted service. The root filesystem (required for CoreOS) will be provided
+by [coreos-downloader][coreos-downloader], similarly to the *standard* flow.
+
+#### Coordination between agents
+
+The first agent to start will be the Ironic agent since it is responsible
+for coordinating the whole flow and (indirectly through Ironic) updating the
+`BareMetalHost`'s status. *Inspection* and additional configuration (RAID etc)
+will be handled by the Ironic agent alone.
+
+Currently, the installation is initiated by applying the following change to
+the `BareMetalHost`'s spec:
+
+```yaml
+Image:
+  URL: url/of/assisted/agent.iso
+  DiskFormat: live-iso
+```
+
+The new process will be instead triggered by:
+
+```yaml
+CustomDeploy:
+  Method: install_assisted
+```
+
+This instructs Ironic to find and execute a *custom deploy step* called
+`install_assisted`, which will be provided by our Ironic agent plugin. This
+step will use D-Bus to order systemd to start the assisted agent and watch
+its status.
+
+If the assisted agent exits successfully, the Ironic agent will report success
+to Ironic and baremetal-operator. Otherwise, a failure will be recorded.
+
+### Risks and Mitigations
+
+- More containers will be downloaded on the ramdisk start-up. We'll need to
+  account for more RAM and a longer start-up time.
+- A user with editing access to `InfraEnv` resources can potentially break
+  the Ironic agent configuration. Short-term, we'll probably document this
+  as a known issue. Long-term, we may need to update `InfraEnv` with a new
+  field.
+
+## Design Details
+
+### Open Questions
+
+- ICC will store the user's pull secret directly to the Ignition override
+  field the `InfraEnv` resource (as a base64 encoded text, not as a reference
+  to secret). Is it acceptable?
```

Comment:
I suspect assisted does it in code. Since it does not understand what our Ignition is doing, it has no way to inject it there.

Maybe we can start the ironic agent in a way that reuses the secret. But it would mean coupling our code with the way the assisted agent is started. Dunno. @flaper87?

Reply:
assisted-service passes it in the file `/root/.docker/config.json` (which is where `podman login` would save it): https://github.com/openshift/assisted-service/blob/master/internal/ignition/ignition.go#L240-L248
i-c-c passes it in the file `/etc/authfile.json`: https://github.com/openshift/image-customization-controller/blob/main/pkg/ignition/service_config.go#L58-L62
All we have to do is use the same file for both.

Reply:
What @zaneb said! Assisted passes the PS as part of the image itself. We need to make sure to use the same for both.

Reply:
The location `/root/.docker/config.json` seems standard. Does it mean that `podman pull` will just work without an explicit authfile argument, @flaper87?

Reply:
Yes.